packages/devtools_app_shared/lib/src/service/dtd_manager.dart

```diff
@@ -18,39 +23,205 @@ class DTDManager {
+        return await connectDtdImpl(uri);
+      } catch (e, s) {
+        // On last attempt, fail and propagate the error.
+        if (attempt == maxRetries) {
+          _connectionState.value = ConnectionFailedDTDState();
+          _log.severe('Failed to connect to DTD after $attempt attempts', e, s);
+          rethrow;
+        }
+
+        // Otherwise, retry after a delay.
+        var delay = math.pow(2, attempt - 1).toInt();
+        _log.info(
+          'Failed to connect to DTD on attempt $attempt, '
+          'will retry in ${delay}s',
+        );
+        while (delay > 0) {
+          _connectionState.value = WaitingToRetryDTDState(delay);
+          await Future.delayed(const Duration(seconds: 1));
+          delay--;
+        }
+      }
+    }
+
+    // We can't get here because of the logic above, but the analyzer can't
+    // tell that.
+    _connectionState.value = NotConnectedDTDState();
+    throw StateError('Failed to connect to DTD');
+  }
+
+  /// Connects Dart Tooling Daemon connection to [uri].
+  ///
+  /// Before connecting to [uri], unless [disconnectBeforeConnecting] is
+  /// `false`, will call [disconnect] to disconnect any existing connection.
+  ///
+  /// If the connection fails, will retry with exponential backoff up to
+  /// [maxRetries].
+  Future<void> _connectImpl(
     Uri uri, {
     void Function(Object, StackTrace?)? onError,
+    int maxRetries = 5,
+    bool disconnectBeforeConnecting = true,
   }) async {
-    await disconnect();
+    if (disconnectBeforeConnecting) {
+      await disconnect();
+    }
+    // Enable automatic reconnect on any new connection.
+    _automaticallyReconnect = true;
 
     try {
-      final connection = await DartToolingDaemon.connect(uri);
+      final connection = await _connectWithRetries(uri, maxRetries: maxRetries);
+
       _uri = uri;
       // Set this after setting the value of [_uri] so that [_uri] can be used
       // by any listeners of the [_connection] notifier.
       _connection.value = connection;
+      _connectionState.value = ConnectedDTDState();
       _log.info('Successfully connected to DTD at: $uri');
+
+      // If a connection drops (and we hadn't disabled auto-reconnect, such
+      // as by explicitly calling disconnect/dispose), we should attempt to
+      // reconnect.
+      unawaited(connection.done.then((_) async {
+        // Trigger disconnect to ensure we emit a `null` connection to
+        // listeners.
+        await disconnectImpl(disableReconnect: false);
+        if (_automaticallyReconnect) {
+          await _connectImpl(
+            uri,
+            onError: onError,
+            // We've already disconnected above, in a way that doesn't disable
+            // reconnect and does not set connection to null (allowing screens
+            // to remain visible under connection overlays).
+            disconnectBeforeConnecting: false,
+          );
+        }
+      }));
```

Comment:
Could we pull everything in `.then` into a named method?

Also consider using `safeUnawaited` (here and elsewhere): https://github.com/flutter/devtools/blob/36af80ea9379f1f7a445e11264a48889bba3eaa9/packages/devtools_app/lib/src/shared/utils/utils.dart#L361-L369

Reply:
Extracted to a method `_reconnectAfterDroppedConnection`.

> Also consider using `safeUnawaited` (here and elsewhere):

It looks like this is defined in `devtools_app` so I don't think it can be used here. Can/should we move it to `devtools_app_shared`?

Reply:
Do you mean just an empty handler? If so, done. If not, what should I do?

I could only find this one `unawaited` added in this PR, but lmk if you think I missed any.

Reply:
Empty handler is fine, thanks!